{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "from __future__ import division\n",
    "\n",
    "from pathlib import Path\n",
    "import argparse\n",
    "import math\n",
    "import numpy as np\n",
    "import os.path as osp\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ROOT_DIR = osp.abspath('.')\n",
    "if ROOT_DIR not in sys.path:\n",
    "    sys.path.append(ROOT_DIR)\n",
    "\n",
    "from utils import io as uio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_intrinsic(filepath, width, height):\n",
    "    import open3d as o3d\n",
    "\n",
    "    m = np.loadtxt(filepath, dtype=np.float32)\n",
    "    intrinsic = o3d.camera.PinholeCameraIntrinsic(width, height, m[0, 0], m[1, 1], m[0, 2], m[1, 2])\n",
    "    return intrinsic\n",
    "\n",
    "def read_extrinsic(filepath):\n",
    "    m = np.loadtxt(filepath, dtype=np.float32)\n",
    "    if np.isnan(m).any():\n",
    "        return None\n",
    "    return m  # (4, 4)\n",
    "\n",
    "\n",
    "def read_rgbd_image(cfg, color_file, depth_file, convert_rgb_to_intensity):\n",
    "    import open3d as o3d\n",
    "    if color_file is None:\n",
    "        color_file = depth_file # to avoid \"Unsupported image format.\"\n",
    "        # rgbd_image = o3d.RGBDImage()\n",
    "        # rgbd_image.depth = o3d.io.read_image(depth_file)\n",
    "        # return rgbd_image\n",
    "    color = o3d.io.read_image(color_file)\n",
    "    depth = o3d.io.read_image(depth_file)\n",
    "    rgbd_image = o3d.geometry.create_rgbd_image_from_color_and_depth(color, depth, cfg.depth_scale, cfg.depth_trunc,\n",
    "                                                                     convert_rgb_to_intensity)\n",
    "    return rgbd_image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_single_fragment(cfg, color_files, depth_files, frag_id, n_frags, intrinsic_path, out_folder):\n",
    "    import open3d as o3d\n",
    "\n",
    "    depth_only_flag = (len(color_files) == 0)\n",
    "    n_frames = len(depth_files)\n",
    "    intrinsic = read_intrinsic(intrinsic_path, cfg.width, cfg.height)\n",
    "    if depth_only_flag:\n",
    "        color_type = o3d.integration.TSDFVolumeColorType.__dict__['None']\n",
    "    else:\n",
    "        color_type = o3d.integration.TSDFVolumeColorType.__dict__['RGB8']\n",
    "        \n",
    "    volume = o3d.integration.ScalableTSDFVolume(voxel_length=cfg.tsdf_cubic_size / 512.0,\n",
    "                                                sdf_trunc=0.04,\n",
    "                                                color_type=color_type)\n",
    "\n",
    "    sid = frag_id * cfg.frames_per_frag\n",
    "    eid = min(sid + cfg.frames_per_frag, n_frames)\n",
    "    pose_base2world = None\n",
    "    pose_base2world_inv = None\n",
    "    for fid in range(sid, eid):\n",
    "        if not depth_only_flag:\n",
    "            color_path = color_files[fid]\n",
    "        else:\n",
    "            color_path = None\n",
    "        depth_path = depth_files[fid]\n",
    "        pose_path = depth_path[:-10] + '.pose.txt'\n",
    "\n",
    "        pose_cam2world = read_extrinsic(pose_path)\n",
    "        if pose_cam2world is None:\n",
    "            continue\n",
    "        if fid == sid:  # Use as base frame\n",
    "            pose_base2world = pose_cam2world\n",
    "            pose_base2world_inv = np.linalg.inv(pose_base2world)\n",
    "        if pose_base2world_inv is None:\n",
    "            break\n",
    "        # Relative camera pose\n",
    "        pose_cam2world = np.matmul(pose_base2world_inv, pose_cam2world)\n",
    "\n",
    "        rgbd = read_rgbd_image(cfg, color_path, depth_path, False)\n",
    "        volume.integrate(rgbd, intrinsic, np.linalg.inv(pose_cam2world))\n",
    "    if pose_base2world_inv is None:\n",
    "        return\n",
    "\n",
    "    pcloud = volume.extract_point_cloud()\n",
    "    o3d.geometry.estimate_normals(pcloud)\n",
    "    o3d.write_point_cloud(osp.join(out_folder, 'cloud_bin_{}.ply'.format(frag_id)), pcloud)\n",
    "\n",
    "    np.save(osp.join(out_folder, 'cloud_bin_{}.pose.npy'.format(frag_id)), pose_base2world)\n",
    "\n",
    "\n",
    "# ---------------------------------------------------------------------------- #\n",
    "# Iterate Folders\n",
    "# ---------------------------------------------------------------------------- #\n",
    "def run_seq(cfg, scene, seq):\n",
    "    print(\"    Start {}\".format(seq))\n",
    "\n",
    "    seq_folder = osp.join(cfg.dataset_root, scene, seq)\n",
    "    color_names = uio.list_files(seq_folder, '*.color.png')\n",
    "    color_paths = [osp.join(seq_folder, cf) for cf in color_names]\n",
    "    depth_names = uio.list_files(seq_folder, '*.depth.png')\n",
    "    depth_paths = [osp.join(seq_folder, df) for df in depth_names]\n",
    "    # depth_paths = [osp.join(seq_folder, cf[:-10] + '.depth.png') for cf in depth_names]\n",
    "\n",
    "    # n_frames = len(color_paths)\n",
    "    n_frames = len(depth_paths)\n",
    "    n_frags = int(math.ceil(float(n_frames) / cfg.frames_per_frag))\n",
    "\n",
    "    out_folder = osp.join(cfg.out_root, scene, seq)\n",
    "    uio.may_create_folder(out_folder)\n",
    "\n",
    "    intrinsic_path = osp.join(cfg.dataset_root, scene, 'camera-intrinsics.txt')\n",
    "\n",
    "    if cfg.threads > 1:\n",
    "        from joblib import Parallel, delayed\n",
    "        import multiprocessing\n",
    "\n",
    "        Parallel(n_jobs=cfg.threads)(\n",
    "            delayed(process_single_fragment)(cfg, color_paths, depth_paths, frag_id, n_frags, intrinsic_path, out_folder)\n",
    "            for frag_id in range(n_frags))\n",
    "\n",
    "    else:\n",
    "        for frag_id in range(n_frags):\n",
    "            process_single_fragment(cfg, color_paths, depth_paths, frag_id, n_frags, intrinsic_path, out_folder)\n",
    "\n",
    "    print(\"    Finished {}\".format(seq))\n",
    "\n",
    "\n",
    "def run_scene(cfg, scene):\n",
    "    print(\"  Start scene {} \".format(scene))\n",
    "\n",
    "    scene_folder = osp.join(cfg.dataset_root, scene)\n",
    "    seqs = uio.list_folders(scene_folder)\n",
    "    print(\"  {} sequences\".format(len(seqs)))\n",
    "    for seq in seqs:\n",
    "        run_seq(cfg, scene, seq)\n",
    "\n",
    "    print(\"  Finished scene {} \".format(scene))\n",
    "\n",
    "\n",
    "def run(cfg):\n",
    "    print(\"Start making fragments\")\n",
    "\n",
    "    uio.may_create_folder(cfg.out_root)\n",
    "\n",
    "    scenes = uio.list_folders(cfg.dataset_root, sort=False)\n",
    "    print(\"{} scenes\".format(len(scenes)))\n",
    "    for scene in scenes:\n",
    "        # if not scene.startswith('analysis'):\n",
    "        #    continue\n",
    "        run_scene(cfg, scene)\n",
    "\n",
    "    print(\"Finished making fragments\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_args():\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument('--dataset_root', default='data/sun3d-hotel_sf-scan1/')\n",
    "    parser.add_argument('--out_root', default='data/3DMatch/rgbd_fragments/')\n",
    "    parser.add_argument('--depth_scale', type=float, default=1000.0)\n",
    "    parser.add_argument('--depth_trunc', type=float, default=6.0)\n",
    "    parser.add_argument('--frames_per_frag', type=int, default=50)\n",
    "    parser.add_argument('--height', type=int, default=480)\n",
    "    parser.add_argument('--threads', type=int, default=8)\n",
    "    parser.add_argument('--tsdf_cubic_size', type=float, default=3.0)\n",
    "    parser.add_argument('--width', type=int, default=640)\n",
    "\n",
    "    return parser.parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] [--dataset_root DATASET_ROOT]\n",
      "                             [--out_root OUT_ROOT] [--depth_scale DEPTH_SCALE]\n",
      "                             [--depth_trunc DEPTH_TRUNC]\n",
      "                             [--frames_per_frag FRAMES_PER_FRAG]\n",
      "                             [--height HEIGHT] [--threads THREADS]\n",
      "                             [--tsdf_cubic_size TSDF_CUBIC_SIZE]\n",
      "                             [--width WIDTH]\n",
      "ipykernel_launcher.py: error: argument --frames_per_frag: invalid int value: 'C:\\\\Users\\\\Darshana\\\\AppData\\\\Local\\\\Temp\\\\tmp-2616rD2Oi7XHMn86.json'\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 2409, in _get_value\n",
      "    result = type_func(arg_string)\n",
      "ValueError: invalid literal for int() with base 10: 'C:\\\\Users\\\\Darshana\\\\AppData\\\\Local\\\\Temp\\\\tmp-2616rD2Oi7XHMn86.json'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 1787, in parse_known_args\n",
      "    namespace, args = self._parse_known_args(args, namespace)\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 1993, in _parse_known_args\n",
      "    start_index = consume_optional(start_index)\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 1933, in consume_optional\n",
      "    take_action(action, args, option_string)\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 1845, in take_action\n",
      "    argument_values = self._get_values(action, argument_strings)\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 2376, in _get_values\n",
      "    value = self._get_value(action, arg_string)\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 2422, in _get_value\n",
      "    raise ArgumentError(action, msg % args)\n",
      "argparse.ArgumentError: argument --frames_per_frag: invalid int value: 'C:\\\\Users\\\\Darshana\\\\AppData\\\\Local\\\\Temp\\\\tmp-2616rD2Oi7XHMn86.json'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Darshana\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-7-c216ee149a8c>\", line 1, in <module>\n",
      "    cfg = parse_args()\n",
      "  File \"<ipython-input-6-e6b95be4ea84>\", line 13, in parse_args\n",
      "    return parser.parse_args()\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 1755, in parse_args\n",
      "    args, argv = self.parse_known_args(args, namespace)\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 1794, in parse_known_args\n",
      "    self.error(str(err))\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 2508, in error\n",
      "    self.exit(2, _('%(prog)s: error: %(message)s\\n') % args)\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\argparse.py\", line 2495, in exit\n",
      "    _sys.exit(status)\n",
      "SystemExit: 2\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Darshana\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"C:\\Users\\Darshana\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"C:\\Users\\Darshana\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"C:\\Users\\Darshana\\anaconda3\\envs\\lidar\\lib\\inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "AttributeError: 'tuple' object has no attribute 'tb_frame'\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "object of type 'NoneType' has no len()",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mValueError\u001B[0m                                Traceback (most recent call last)",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36m_get_value\u001B[1;34m(self, action, arg_string)\u001B[0m\n\u001B[0;32m   2408\u001B[0m         \u001B[1;32mtry\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 2409\u001B[1;33m             \u001B[0mresult\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mtype_func\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0marg_string\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   2410\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;31mValueError\u001B[0m: invalid literal for int() with base 10: 'C:\\\\Users\\\\Darshana\\\\AppData\\\\Local\\\\Temp\\\\tmp-2616rD2Oi7XHMn86.json'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001B[1;31mArgumentError\u001B[0m                             Traceback (most recent call last)",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36mparse_known_args\u001B[1;34m(self, args, namespace)\u001B[0m\n\u001B[0;32m   1786\u001B[0m         \u001B[1;32mtry\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 1787\u001B[1;33m             \u001B[0mnamespace\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0margs\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0m_parse_known_args\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0margs\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mnamespace\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   1788\u001B[0m             \u001B[1;32mif\u001B[0m \u001B[0mhasattr\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mnamespace\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0m_UNRECOGNIZED_ARGS_ATTR\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36m_parse_known_args\u001B[1;34m(self, arg_strings, namespace)\u001B[0m\n\u001B[0;32m   1992\u001B[0m             \u001B[1;31m# consume the next optional and any arguments for it\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 1993\u001B[1;33m             \u001B[0mstart_index\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mconsume_optional\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mstart_index\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   1994\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36mconsume_optional\u001B[1;34m(start_index)\u001B[0m\n\u001B[0;32m   1932\u001B[0m             \u001B[1;32mfor\u001B[0m \u001B[0maction\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0margs\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0moption_string\u001B[0m \u001B[1;32min\u001B[0m \u001B[0maction_tuples\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 1933\u001B[1;33m                 \u001B[0mtake_action\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0maction\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0margs\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0moption_string\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   1934\u001B[0m             \u001B[1;32mreturn\u001B[0m \u001B[0mstop\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36mtake_action\u001B[1;34m(action, argument_strings, option_string)\u001B[0m\n\u001B[0;32m   1844\u001B[0m             \u001B[0mseen_actions\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0madd\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0maction\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 1845\u001B[1;33m             \u001B[0margument_values\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0m_get_values\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0maction\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0margument_strings\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   1846\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36m_get_values\u001B[1;34m(self, action, arg_strings)\u001B[0m\n\u001B[0;32m   2375\u001B[0m             \u001B[0marg_string\u001B[0m\u001B[1;33m,\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0marg_strings\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 2376\u001B[1;33m             \u001B[0mvalue\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0m_get_value\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0maction\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0marg_string\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   2377\u001B[0m             \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0m_check_value\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0maction\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mvalue\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36m_get_value\u001B[1;34m(self, action, arg_string)\u001B[0m\n\u001B[0;32m   2421\u001B[0m             \u001B[0mmsg\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0m_\u001B[0m\u001B[1;33m(\u001B[0m\u001B[1;34m'invalid %(type)s value: %(value)r'\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 2422\u001B[1;33m             \u001B[1;32mraise\u001B[0m \u001B[0mArgumentError\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0maction\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mmsg\u001B[0m \u001B[1;33m%\u001B[0m \u001B[0margs\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   2423\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;31mArgumentError\u001B[0m: argument --frames_per_frag: invalid int value: 'C:\\\\Users\\\\Darshana\\\\AppData\\\\Local\\\\Temp\\\\tmp-2616rD2Oi7XHMn86.json'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001B[1;31mSystemExit\u001B[0m                                Traceback (most recent call last)",
      "    \u001B[1;31m[... skipping hidden 1 frame]\u001B[0m\n",
      "\u001B[1;32m<ipython-input-7-c216ee149a8c>\u001B[0m in \u001B[0;36m<module>\u001B[1;34m\u001B[0m\n\u001B[1;32m----> 1\u001B[1;33m \u001B[0mcfg\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mparse_args\u001B[0m\u001B[1;33m(\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m",
      "\u001B[1;32m<ipython-input-6-e6b95be4ea84>\u001B[0m in \u001B[0;36mparse_args\u001B[1;34m()\u001B[0m\n\u001B[0;32m     12\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m---> 13\u001B[1;33m     \u001B[1;32mreturn\u001B[0m \u001B[0mparser\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mparse_args\u001B[0m\u001B[1;33m(\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36mparse_args\u001B[1;34m(self, args, namespace)\u001B[0m\n\u001B[0;32m   1754\u001B[0m     \u001B[1;32mdef\u001B[0m \u001B[0mparse_args\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mself\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0margs\u001B[0m\u001B[1;33m=\u001B[0m\u001B[1;32mNone\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mnamespace\u001B[0m\u001B[1;33m=\u001B[0m\u001B[1;32mNone\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 1755\u001B[1;33m         \u001B[0margs\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0margv\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mparse_known_args\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0margs\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mnamespace\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   1756\u001B[0m         \u001B[1;32mif\u001B[0m \u001B[0margv\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36mparse_known_args\u001B[1;34m(self, args, namespace)\u001B[0m\n\u001B[0;32m   1793\u001B[0m             \u001B[0merr\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0m_sys\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mexc_info\u001B[0m\u001B[1;33m(\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m[\u001B[0m\u001B[1;36m1\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 1794\u001B[1;33m             \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0merror\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mstr\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0merr\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   1795\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36merror\u001B[1;34m(self, message)\u001B[0m\n\u001B[0;32m   2507\u001B[0m         \u001B[0margs\u001B[0m \u001B[1;33m=\u001B[0m \u001B[1;33m{\u001B[0m\u001B[1;34m'prog'\u001B[0m\u001B[1;33m:\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mprog\u001B[0m\u001B[1;33m,\u001B[0m \u001B[1;34m'message'\u001B[0m\u001B[1;33m:\u001B[0m \u001B[0mmessage\u001B[0m\u001B[1;33m}\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 2508\u001B[1;33m         \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mexit\u001B[0m\u001B[1;33m(\u001B[0m\u001B[1;36m2\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0m_\u001B[0m\u001B[1;33m(\u001B[0m\u001B[1;34m'%(prog)s: error: %(message)s\\n'\u001B[0m\u001B[1;33m)\u001B[0m \u001B[1;33m%\u001B[0m \u001B[0margs\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m",
      "\u001B[1;32m~\\anaconda3\\envs\\lidar\\lib\\argparse.py\u001B[0m in \u001B[0;36mexit\u001B[1;34m(self, status, message)\u001B[0m\n\u001B[0;32m   2494\u001B[0m             \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0m_print_message\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mmessage\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0m_sys\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mstderr\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 2495\u001B[1;33m         \u001B[0m_sys\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mexit\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mstatus\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   2496\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;31mSystemExit\u001B[0m: 2",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001B[1;31mTypeError\u001B[0m                                 Traceback (most recent call last)",
      "    \u001B[1;31m[... skipping hidden 1 frame]\u001B[0m\n",
      "\u001B[1;32m~\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\interactiveshell.py\u001B[0m in \u001B[0;36mshowtraceback\u001B[1;34m(self, exc_tuple, filename, tb_offset, exception_only, running_compiled_code)\u001B[0m\n\u001B[0;32m   2053\u001B[0m                            'the full traceback.\\n']\n\u001B[0;32m   2054\u001B[0m                     stb.extend(self.InteractiveTB.get_exception_only(etype,\n\u001B[1;32m-> 2055\u001B[1;33m                                                                      value))\n\u001B[0m\u001B[0;32m   2056\u001B[0m                 \u001B[1;32melse\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m   2057\u001B[0m                     \u001B[1;32mtry\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\ultratb.py\u001B[0m in \u001B[0;36mget_exception_only\u001B[1;34m(self, etype, value)\u001B[0m\n\u001B[0;32m    752\u001B[0m         \u001B[0mvalue\u001B[0m \u001B[1;33m:\u001B[0m \u001B[0mexception\u001B[0m \u001B[0mvalue\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    753\u001B[0m         \"\"\"\n\u001B[1;32m--> 754\u001B[1;33m         \u001B[1;32mreturn\u001B[0m \u001B[0mListTB\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mstructured_traceback\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mself\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0metype\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mvalue\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m    755\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    756\u001B[0m     \u001B[1;32mdef\u001B[0m \u001B[0mshow_exception_only\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mself\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0metype\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mevalue\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\ultratb.py\u001B[0m in \u001B[0;36mstructured_traceback\u001B[1;34m(self, etype, evalue, etb, tb_offset, context)\u001B[0m\n\u001B[0;32m    631\u001B[0m                     chained_exceptions_tb_offset, context)\n\u001B[0;32m    632\u001B[0m                 \u001B[1;33m+\u001B[0m \u001B[0mchained_exception_message\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m--> 633\u001B[1;33m                 + out_list)\n\u001B[0m\u001B[0;32m    634\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    635\u001B[0m         \u001B[1;32mreturn\u001B[0m \u001B[0mout_list\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\ultratb.py\u001B[0m in \u001B[0;36mstructured_traceback\u001B[1;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001B[0m\n\u001B[0;32m   1366\u001B[0m             \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mtb\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mtb\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m   1367\u001B[0m         return FormattedTB.structured_traceback(\n\u001B[1;32m-> 1368\u001B[1;33m             self, etype, value, tb, tb_offset, number_of_lines_of_context)\n\u001B[0m\u001B[0;32m   1369\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m   1370\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\ultratb.py\u001B[0m in \u001B[0;36mstructured_traceback\u001B[1;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001B[0m\n\u001B[0;32m   1266\u001B[0m             \u001B[1;31m# Verbose modes need a full traceback\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m   1267\u001B[0m             return VerboseTB.structured_traceback(\n\u001B[1;32m-> 1268\u001B[1;33m                 \u001B[0mself\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0metype\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mvalue\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mtb\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mtb_offset\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mnumber_of_lines_of_context\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   1269\u001B[0m             )\n\u001B[0;32m   1270\u001B[0m         \u001B[1;32melif\u001B[0m \u001B[0mmode\u001B[0m \u001B[1;33m==\u001B[0m \u001B[1;34m'Minimal'\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\ultratb.py\u001B[0m in \u001B[0;36mstructured_traceback\u001B[1;34m(self, etype, evalue, etb, tb_offset, number_of_lines_of_context)\u001B[0m\n\u001B[0;32m   1123\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m   1124\u001B[0m         formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n\u001B[1;32m-> 1125\u001B[1;33m                                                                tb_offset)\n\u001B[0m\u001B[0;32m   1126\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m   1127\u001B[0m         \u001B[0mcolors\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mColors\u001B[0m  \u001B[1;31m# just a shorthand + quicker name lookup\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\ultratb.py\u001B[0m in \u001B[0;36mformat_exception_as_a_whole\u001B[1;34m(self, etype, evalue, etb, number_of_lines_of_context, tb_offset)\u001B[0m\n\u001B[0;32m   1080\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m   1081\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m-> 1082\u001B[1;33m         \u001B[0mlast_unique\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mrecursion_repeat\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mfind_recursion\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0morig_etype\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mevalue\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mrecords\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m   1083\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m   1084\u001B[0m         \u001B[0mframes\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mformat_records\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mrecords\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mlast_unique\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mrecursion_repeat\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32m~\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\ultratb.py\u001B[0m in \u001B[0;36mfind_recursion\u001B[1;34m(etype, value, records)\u001B[0m\n\u001B[0;32m    380\u001B[0m     \u001B[1;31m# first frame (from in to out) that looks different.\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    381\u001B[0m     \u001B[1;32mif\u001B[0m \u001B[1;32mnot\u001B[0m \u001B[0mis_recursion_error\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0metype\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mvalue\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mrecords\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m--> 382\u001B[1;33m         \u001B[1;32mreturn\u001B[0m \u001B[0mlen\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mrecords\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m,\u001B[0m \u001B[1;36m0\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m    383\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    384\u001B[0m     \u001B[1;31m# Select filename, lineno, func_name to track frames with\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;31mTypeError\u001B[0m: object of type 'NoneType' has no len()"
     ]
    }
   ],
   "source": [
    "cfg = parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "_StoreAction(option_strings=['--width'], dest='width', nargs=None, const=None, default=640, type=<class 'int'>, choices=None, help=None, metavar=None)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument('--dataset_root', default='data/sun3d-hotel_sf-scan1/')\n",
    "parser.add_argument('--out_root', default='data/3DMatch/rgbd_fragments/')\n",
    "parser.add_argument('--depth_scale', type=float, default=1000.0)\n",
    "parser.add_argument('--depth_trunc', type=float, default=6.0)\n",
    "# parser.add_argument('--frames_per_frag', type=int, default=50)\n",
    "parser.add_argument('--height', type=int, default=480)\n",
    "parser.add_argument('--threads', type=int, default=8)\n",
    "parser.add_argument('--tsdf_cubic_size', type=float, default=3.0)\n",
    "parser.add_argument('--width', type=int, default=640)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] [--dataset_root DATASET_ROOT]\n",
      "                             [--out_root OUT_ROOT] [--depth_scale DEPTH_SCALE]\n",
      "                             [--depth_trunc DEPTH_TRUNC] [--height HEIGHT]\n",
      "                             [--threads THREADS]\n",
      "                             [--tsdf_cubic_size TSDF_CUBIC_SIZE]\n",
      "                             [--width WIDTH]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: --ip=127.0.0.1 --stdin=9003 --control=9001 --hb=9000 --Session.signature_scheme=\"hmac-sha256\" --Session.key=b\"99e0c9d4-7c59-482b-be09-04611c31a8b4\" --shell=9002 --transport=\"tcp\" --iopub=9004 --f=C:\\Users\\Darshana\\AppData\\Local\\Temp\\tmp-2616rD2Oi7XHMn86.json\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001B[1;31mSystemExit\u001B[0m\u001B[1;31m:\u001B[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Darshana\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\core\\interactiveshell.py:3445: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "parser.parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "5e21840724f3defa735c04c75029b670fafa8635d8c12fcfe3eb8a6ab8ff3791"
  },
  "kernelspec": {
   "name": "python3",
   "language": "python",
   "display_name": "Python 3 (ipykernel)"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}